substrate-node/pallets/pallet-tfgrid/src/grid_migration.rs

```diff
@@ -0,0 +1,263 @@
+                    country: BoundedVec::try_from(node::DEFAULT_COUNTRY_NAME.to_vec()).unwrap(),
+                    latitude: BoundedVec::try_from(node::DEFAULT_LATITUDE.to_vec()).unwrap(),
+                    longitude: BoundedVec::try_from(node::DEFAULT_LONGITUDE.to_vec()).unwrap(),
+                };
+                info!("set default location for node");
+                <T as Config>::Location::try_from(default_location_input).unwrap()
+            }
+        };
+
+        let serial_number_input: SerialNumberInput =
+            BoundedVec::try_from(node.serial_number.clone()).unwrap();
+        let serial_number = match <T as Config>::SerialNumber::try_from(serial_number_input) {
+            Ok(serial) => serial,
+            Err(e) => {
+                info!(
+                    "serial number: {:?}",
+                    core::str::from_utf8(&node.serial_number).unwrap()
+                );
+                info!(
+                    "failed to parse serial number for node: {:?}, error: {:?}",
+                    k, e
+                );
+                let default_serial_number_input: SerialNumberInput =
+                    BoundedVec::try_from(node::DEFAULT_SERIAL_NUMBER.to_vec()).unwrap();
+                info!("set default serial number for node");
+                <T as Config>::SerialNumber::try_from(default_serial_number_input).unwrap()
+            }
+        };
+
+        let new_node = TfgridNode::<T> {
+            version: 6, // ??
+            id: node.id,
+            farm_id: node.farm_id,
+            twin_id: node.twin_id,
+            resources: node.resources,
+            location,
+            public_config: node.public_config,
+            created: node.created,
+            farming_policy_id: node.farming_policy_id,
+            interfaces: node.interfaces,
+            certification: node.certification,
+            secure_boot: node.secure_boot,
+            virtualized: node.virtualized,
+            serial_number,
+            connection_price: node.connection_price,
+        };
+
+        migrated_count += 1;
+
+        Some(new_node)
+    });
+    info!(
+        " <<< Node storage updated! Migrated {} nodes ✅",
+        migrated_count
+    );
+
+    // Return the weight consumed by the migration.
+    T::DbWeight::get().reads_writes(migrated_count as Weight + 1, migrated_count as Weight + 1)
+}
+
+fn update_pallet_storage_version<T: Config>() -> frame_support::weights::Weight {
+    PalletVersion::<T>::set(types::StorageVersion::V10Struct);
+    info!(" <<< Storage version upgraded");
+
+    // Return the weight consumed by the migration.
+    T::DbWeight::get().writes(1)
+}
+
+fn get_location<T: Config>(
+    node: &deprecated::Node<PubConfigOf<T>, InterfaceOf<T>>,
+) -> Result<LocationOf<T>, Error<T>> {
+    let location_input = LocationInput {
+        city: BoundedVec::try_from(node.city.clone()).unwrap(),
```

Comment:
Do you really need to specify the input type first?

Reply:
I think yes because it is how the old location goes through the validation process 
`location_input` is parsed in `<T as Config>::Location::try_from(location_input)`
If the validation process throws error then old location is replaced by default location